### PR TITLE
[feature/diagnostic] Diagnostic information

### PR DESCRIPTION
**Pull request by felix-schwarz**
## Description
This PR aims to surface diagnostic information and actions from the SDK and make it available inline, for export and for logging.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Screenshots
| Row action | Diagnostic Info |
| -- | -- |
| Bildschirmfoto 2020-07-28 um 00 54 44 | Bildschirmfoto 2020-07-28 um 00 54 48 |

| Diagnostic Overview (in Account Management) | Changed Status label | Info for Fetching Updates |
| -- | -- | -- |
| Screen Shot 2020-08-01 at 01 04 42 | Screen Shot 2020-08-01 at 01 04 31 | Screen Shot 2020-08-01 at 01 05 03 |

## Sample Markdown/HTML report

This is an example report as exported by the "Share" button inside the Diagnostic viewer:

---
# Sync Record 2
<table>
<tr><td nowrap valign="top">Sync Record ID</td><td><code>2</code></td></tr>
<tr><td nowrap valign="top">State</td><td><code>2</code></td></tr>
<tr><td nowrap valign="top">In progress since</td><td><code>2020-07-29 07:59:58 +0000</code></td></tr>
<tr><td nowrap valign="top">Origin Process</td><td><code>&lt;OCProcessSession: 0x7fc868fc2ec0, uuid: 089B8384-614A-46E3-8038-A26682AA856C, bundleIdentifier: com.owncloud.ios-app.ownCloud-File-Provider, lastActive: 2020-07-29 07:59:41 +0000, bootTimestamp: 1595789443, processType: 1&gt;</code></td></tr>
<tr><td nowrap valign="top">Origin Process Valid</td><td><code>0</code></td></tr>
<tr><td nowrap valign="top">Timestamp</td><td><code>2020-07-29 07:59:58 +0000</code></td></tr>
<tr><td nowrap valign="top">Lane ID</td><td><code>2</code></td></tr>
<tr><td nowrap valign="top">Local ID</td><td><code>15C5C996041F4ED1B23463A861B34CE0</code></td></tr>
<tr><td nowrap valign="top">Action ID</td><td><code>createFolder</code></td></tr>
<tr><td nowrap valign="top">Action</td><td>
<table>
<tr><td nowrap valign="top">Class</td><td><code>OCSyncActionCreateFolder</code></td></tr>
<tr><td nowrap valign="top">Identifier</td><td><code>createFolder</code></td></tr>
<tr><td nowrap valign="top">Localized Description</td><td><code>Creating folder ownCloud Manual.pdf…</code></td></tr>
</table>

</td></tr>
<tr><td nowrap valign="top">Wait Conditions</td><td>
<table>
<tr><td nowrap valign="top"># 1</td><td>
<table>
<tr><td nowrap valign="top">Type</td><td><code>OCWaitConditionIssue</code></td></tr>
<tr><td nowrap valign="top">UUID</td><td><code>7216ED7F-380C-4303-B9C9-20AC83400D37</code></td></tr>
<tr><td nowrap valign="top">Issue</td><td><code>&lt;OCSyncIssue: 0x7fc818c06420, uuid: 7C461517-F145-4D1E-B002-6C3F6B4E91F9, syncRecordID: 2, templateIdentifier: createFolder._cancel.nonDestructive, level: 2, title: Couldn't create ownCloud Manual.pdf, description: The resource you tried to create already exists, choices: (
    "&lt;OCSyncIssueChoice: 0x7fc818c1ab00, identifier: _cancel, autoChoiceForError: (null), label: Cancel&gt;"
)&gt;</code></td></tr>
<tr><td nowrap valign="top">ProcessSession</td><td><code>&lt;OCProcessSession: 0x7fc818c3f180, uuid: FEE52540-6070-4132-AD06-9BBF229E0E89, bundleIdentifier: com.owncloud.ios-app, lastActive: 2020-07-29 08:40:56 +0000, bootTimestamp: 1595789443, processType: 0&gt;</code></td></tr>
<tr><td nowrap valign="top">Resolved</td><td><code>0</code></td></tr>
</table>

</td></tr>
</table>

</td></tr>
<tr><td nowrap valign="top">HTTP Requests</td><td>
<table>
<tr><td nowrap valign="top">ephermal</td><td>
<table>
</table>

</td></tr>
<tr><td nowrap valign="top">default</td><td>
<table>
</table>

</td></tr>
</table>

</td></tr>
</table>

### QA

- [x] (1) Number of folders includes root https://github.com/owncloud/ios-app/pull/762#issuecomment-671990020 [FIXED]
- [x] (2) `Http Requests` not reachable https://github.com/owncloud/ios-app/pull/762#issuecomment-671991620 [FIXED]
- [ ] (3) `Process Sync Records `https://github.com/owncloud/ios-app/pull/762#issuecomment-672004696

### CODE REVIEW

File: ownCloud/Diagnostic/DiagnosticViewController.swift [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	init(for node: OCDiagnosticNode, context: OCDiagnosticContext?, style: Style = .hierarchical) {
+		self.context = context
+		self.style = style
+
+		super.init(style: .grouped)
+
+		self.rootNode = node
+
+		self.navigationItem.title = node.label
+		self.navigationItem.rightBarButtonItem = UIBarButtonItem(image: UIImage(named: "open-in"), style: .plain, target: self, action: #selector(self.shareAsMarkdown(_:)))
```
Comment (hosy) [inline ownCloud/Diagnostic/DiagnosticViewController.swift:52]: Would be nice to have an `accessibilityLabel`, when the `UIBarButtonItem` is image only

**Comment by jesmrec** [comment]
In QA, i will not enter in the containing information in the sync reports.

**Comment by jesmrec** [comment]
### (1)

The account skeleton contains 2 folders (`Documents`, `Photos`) and 5 files (`ownCloud Manual`, 3 pics inside `Photos` and one file inside `Documents`).

The Diagnostic overview shows 3 folders (does it include root folder ?):

Screenshot 2020-08-11 at 16 39 24

**Comment by jesmrec** [comment]
### (2)

The option in the bottom of Sync Record `HTTP Requests` is not reachable. An scroll would be helpful here:

Aug-11-2020 16-29-35

**Comment by jesmrec** [comment]
### (3)

How can i check the option `Process Sync Records`?. I caused different errors, but nothing happened when i clicked such option

**Comment by hosy** [comment]
@jesmrec I fixed finding (1) and (2). Please give it a new try.
@felix-schwarz please note that (1) and (2) is already fixed.

**Comment by jesmrec** [comment]
(1) and (2) are fixed

(3) is a question, just in case something is not correct. At least for me, i couldn't reach any information in `Process Sync Records`

**Comment by hosy** [comment]
@jesmrec regarding (3): I added a notification banner, when a diagnostic action was performed to give the user feedback. As I can see, this will only be collected, if available.

**Comment by jesmrec** [comment]
Ok, i see the notification. 

From my side you can move this forward. As i commented above, the content of sync records and other inline information is not valuable from my side, this is something useful for you. Feel free to iterate over the feature to get the information you need in every moment.

**Pull request merged**